Intersect.Client/Entities/Player.cs

```diff
@@ -1974,137 +1991,162 @@ private void ProcessDirectionalInput()
-                                Dir = 1;
-                                OffsetY = -Options.TileHeight;
-                                OffsetX = 0;
-                            }
+            if (Options.Combat.MovementCancelsCast)
+            {
+                CastTime = 0;
+            }
 
-                            break;
-                        case 2: // Left
-                            if (IsTileBlocked(X - 1, Y, Z, MapId, ref blockedBy) == -1)
-                            {
-                                tmpX--;
-                                IsMoving = true;
-                                Dir = 2;
-                                OffsetY = 0;
-                                OffsetX = Options.TileWidth;
-                            }
+            var deltaX = 0;
+            var deltaY = 0;
 
-                            break;
-                        case 3: // Right
-                            if (IsTileBlocked(X + 1, Y, Z, MapId, ref blockedBy) == -1)
-                            {
-                                tmpX++;
-                                IsMoving = true;
-                                Dir = 3;
-                                OffsetY = 0;
-                                OffsetX = -Options.TileWidth;
-                            }
+            switch (MoveDir)
+            {
+                case Direction.Up:
+                    deltaX = 0;
+                    deltaY = -1;
+                    break;
+                case Direction.Down:
+                    deltaX = 0;
+                    deltaY = 1;
+                    break;
+                case Direction.Left:
+                    deltaX = -1;
+                    deltaY = 0;
+                    break;
+                case Direction.Right:
+                    deltaX = 1;
+                    deltaY = 0;
+                    break;
+                case Direction.UpLeft:
+                    deltaX = -1;
+                    deltaY = -1;
+                    break;
+                case Direction.UpRight:
+                    deltaX = 1;
+                    deltaY = -1;
+                    break;
+                case Direction.DownLeft:
+                    deltaX = -1;
+                    deltaY = 1;
+                    break;
+                case Direction.DownRight:
+                    deltaX = 1;
+                    deltaY = 1;
+                    break;
+            }
 
-                            break;
+            if (deltaX != 0 || deltaY != 0)
+            {
+                var newX = tmpX + deltaX;
+                var newY = tmpY + deltaY;
+                if (IsTileBlocked(newX, newY, Z, MapId, ref blockedBy) == -1)
+                {
+                    tmpX += (sbyte)deltaX;
+                    tmpY += (sbyte)deltaY;
+                    IsMoving = true;
+                    Dir = MoveDir;
+                    if (deltaX == 0)
+                    {
+                        OffsetX = 0;
```

Comment:
Please see my "skip" comments in the server implementation, this applies to everything down to L2149.

Reply:
```csharp
            var skip = deltaX == 0 && deltaY == 0;

            if (!skip)
            {
                var newX = tmpX + deltaX;
                var newY = tmpY + deltaY;
                var isBlocked = IsTileBlocked(newX, newY, Z, MapId, ref blockedBy) == -1;

                if (isBlocked && newX >= 0 && newX < Options.MapWidth && newY >= 0 && newY < Options.MapHeight)
                {
                    tmpX += (sbyte)deltaX;
                    tmpY += (sbyte)deltaY;
                    IsMoving = true;
                    Dir = MoveDir;

                    if (deltaX == 0)
                    {
                        OffsetX = 0;
                    }
                    else
                    {
                        OffsetX = deltaX > 0 ? -Options.TileWidth : Options.TileWidth;
                    }

                    if (deltaY == 0)
                    {
                        OffsetY = 0;
                    }
                    else
                    {
                        OffsetY = deltaY > 0 ? -Options.TileHeight : Options.TileHeight;
                    }
                }
            }
```